org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenExecutionContext.java

```diff
@@ -92,29 +92,30 @@ public class MavenExecutionContext implements IMavenExecutionContext {
 
   private MavenExecutionRequest request;
 
-  // TODO maybe delegate to parent context
   private Map<String, Object> context;
 
   private final File basedir;
 
-  private IMavenProjectFacade projectFacade;
+  private final Function<? super MavenExecutionContext, MavenProject> projectSupplier;
 
-  private IComponentLookup containerLookup;
+  private final IComponentLookup containerLookup;
 
-  public MavenExecutionContext(IComponentLookup lookup, IMavenProjectFacade projectFacade) {
-    this.containerLookup = lookup;
-    this.projectFacade = projectFacade;
-    if(projectFacade != null) {
-      IFile pom = projectFacade.getPom();
-      if(pom != null && pom.getLocation() != null) {
-        File pomLocation = pom.getLocation().toFile();
-        this.basedir = pomLocation.isDirectory() ? pomLocation : pomLocation.getParentFile();
-      } else {
-        this.basedir = null;
-      }
-    } else {
-      this.basedir = null;
-    }
+  /**
+   * Creates a new execution context for the given environment.
+   * 
+   * @param lookup the lookup that is used to query for components, this lookup must support to lookup at least a
+   *          {@link PlexusContainer} for session setup
+   * @param baseDir the basedir of the execution, this is effectively the folder where one normally would execute the
+   *          maven command, also often refereed to as the working directory of the process. If it is <code>null</code>
+   *          we assume an anonymous execution, some feature might not work in such a context.
+   * @param projectSupplier if an execution is started without an explicit project and the projectSupplier is not
+   *          <code>null</code> it is queried to supply the MavenProject used for session setup.
+   */
+  public MavenExecutionContext(IComponentLookup lookup, File baseDir,
+      Function<? super MavenExecutionContext, MavenProject> projectSupplier) {
```

Comment:
Is there a reason that `projectSupplier` is a Function and not a real supplier or simply passing the `MavenProject` directly?
Because the only caller that does not pass null, just passes a function that always returns the same project.
So I wonder if the last argument could just be a MavenProject?
Or is this required for some later work?